### PR TITLE
Add some dependency logic around generators.

### PR DESCRIPTION
**Pull request by vin**
This isn't 100% correct, but it's a pretty good approximation.
Specifically, it doesn't know about transitive dependencies from other
go packages.  But it'll rebuild a generator if any code inside its
package changes (or build them all if anything in gengo changes).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -135,42 +135,23 @@ $(BINDIR)/service-catalog: .init .generate_files cmd/service-catalog
...
-
-$(BINDIR)/client-gen: .init
-	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/code-generator/cmd/client-gen
+GENERATORS = $(addprefix $(BINDIR)/, defaulter-gen deepcopy-gen conversion-gen \
+	     client-gen lister-gen informer-gen openapi-gen)
 
-$(BINDIR)/lister-gen: .init
-	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/code-generator/cmd/lister-gen
+.PHONY: generators
+generators: $(GENERATORS)
```
Comment (MHBauer) [inline Makefile]: what's this target for? not used anywhere

File: Makefile [hunk 2]
```diff
@@ -135,42 +135,23 @@ $(BINDIR)/service-catalog: .init .generate_files cmd/service-catalog
...
 
-$(BINDIR)/informer-gen: .init
-	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/code-generator/cmd/informer-gen
+.SECONDEXPANSION:
 
-$(BINDIR)/openapi-gen: vendor/k8s.io/code-generator/cmd/openapi-gen
-	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/$^
+$(BINDIR)/%-gen: $$(shell find vendor/k8s.io/code-generator/cmd/$$*-gen vendor/k8s.io/gengo) .init
```
Comment (MHBauer) [inline Makefile]: notes, maybe need some comments to quickstart people's minds when reading 

depend on something-gen in gengo

File: Makefile [hunk 3]
```diff
@@ -135,42 +135,23 @@ $(BINDIR)/service-catalog: .init .generate_files cmd/service-catalog
...
+	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/code-generator/cmd/$*-gen
```
Comment (MHBauer) [inline Makefile]: build the something-gen

**Comment by MHBauer** [comment]
on golang 1.10 and it's handling the binary dependency thing and caching.

File: Makefile [hunk 4]
```diff
@@ -143,42 +143,26 @@ $(BINDIR)/service-catalog: .init .generate_files cmd/service-catalog
...
-	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/code-generator/cmd/informer-gen
+.SECONDEXPANSION:
 
-$(BINDIR)/openapi-gen: vendor/k8s.io/code-generator/cmd/openapi-gen
-	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/$^
+# We specify broad dependencies for these generator binaries: each one depends
+# on everything under its source tree as well as gengo's.  This uses GNU Make's
+# secondary expansion feature to pass $* to `find`.
+$(BINDIR)/%-gen: $$(shell find vendor/k8s.io/code-generator/cmd/$$*-gen vendor/k8s.io/gengo) .init
+	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/code-generator/cmd/$*-gen
```
Comment (MHBauer) [inline Makefile]: nice comment. this is neat. I'm wondering if the find is the right practice vs having golang handle it vs handling it like upstream k/k does with the go2mk generator.

**Comment by MHBauer** [comment]
altogether, I cannot see any reason not to do this.
/lgtm
/approve

**Comment by MHBauer** [comment]
/assign @jberkhahn 
for testing on mac

/assign @jboyd01 
as someone who commented in the meeting

**Comment by jberkhahn** [comment]
this seems to work fine on Mac

/lgtm

**Pull request merged**